platform/engineering/design-docs/2020-04-16-integration-testing-with-consumer-driven-contract-tests.md

```diff
@@ -0,0 +1,275 @@
+- The total run time of the CI pipeline increases since there is another dependency to build, so the tests will be even slower.
+- Setting up the API creates another potential point of failure that isn't relevant to the tests being run. Failures related to environment or configuration while building `vets-api` doesn't add value to the actual tests but does add time invested in maintenance.
+- Introducing `vets-api` to the pipeline adds a dependency that is unfamiliar to the developers writing the tests. FE developers are responsible for maintaining these tests since they are UI-based, and having to manage building and running `vets-api` adds to the maintenance burden.
+- We also need to consider the dependency on Betamocks, since we would want to stub out external services to keep the scope of the tests focused on VA.gov. Modifying the `vets-api-mockdata` repo is unintuitive to most FE engineers because it requires understanding what responses the API should expect from external services. That knowledge is out of scope for FE engineers who will inevitably have to tweak the mocks to accommodate their tests. Alternatively, BE developers could be requested to make the necessary changes to the mocks, but they are not the maintainers of the tests, so they would need to take the time to understand the needs of the FE tests instead of focusing on their own work. In either case, developers are forced to overextend beyond their domain.
+
+With Pact, we get the benefits of both types of testing.
+- The contract tests that generate the pacts are unit-like, so they run quickly and reliably, and failures are easy to isolate because the tests are written against specific endpoints.
+- The pact publishing and verification loop does still require building `vets-api`. However, the same problems with building the API from the full E2E approach are mitigated with the contract testing approach. The workflow cleanly isolates and delineates the responsibilities of FE and BE engineers, so any issues within either FE or BE pipeline can be handled by the respective party. Furthermore, there is no reliance on Betamocks, and the needs of the tests are formalized in the Pact DSL.
+- When run alone, contract tests don't simulate real users. However, we can use the generated pacts to create stubbed responses in the E2E tests. This would also address the existing issue of unvalidated and untrustworthy stubbed responses in those tests. There is already an official stub server that Pact has documented. Alternatively, in the spirit of consolidating dependencies, it might be simpler to leverage the stubbing capabilities of the E2E testing framework. Either way, using the the pacts grants the advantage of having stubbed responses that have gone through validation with the actual running API. The validation may occur in separate steps, but those steps come together to make the contract-supplemented E2E tests as trustworthy as full E2E tests.
```

Comment:
So the idea here is that we could use the end output of this contract test in the functional FE tests?

Reply:
That's correct. I should update the document to talk about why we don't want the functional tests to generate the contracts (Pact has an explicit answer in the FAQ about that), but I see the functional tests setting up the stubs using the end output from the contract tests. That end output will a set of mock responses that we can trust because they've passed validation from the API.